### ISSUE TITLE
Type-safe column definitions

### ISSUE THREAD

**Issue by zewa666** [ghiscoding/aurelia-slickgrid#441]
I was looking at the [Column interface](https://github.com/ghiscoding/aurelia-slickgrid/blob/master/src/aurelia-slickgrid/models/column.interface.ts#L14) and found a generic used. This though is being used for the `dataContext` argument of `asyncPostRenderer`. Now in my opinion this does not make really lots of sense since if not defined otherwise via `params` dataContext is going to be an object with keys named after field names right?

So in order to achieve type-safety for column fields and the renderer here's what I do:

```typescript
type GridFields = "FieldA" | "FieldB" | "FieldC";

export class App {
    public columnDefinitions: (Column & { field: GridFields })[];

    private defineGrid() {
      this.columnDefinitions = [
        { id: "FieldA", name: "Field A", field: "FieldA" /** THIS BECOMES TYPE SAFE **/},
        { id: "FieldB", name: "Field B", field: "FieldB" /** THIS BECOMES TYPE SAFE **/},
        { id: "FieldC", name: "Field C", field: "FieldC" /** THIS BECOMES TYPE SAFE **/,
          asyncPostRender: (cellNode: JQuery<HTMLElement>, row: number, dataContext: {[key in GridFields]: unknown}, colDef: Column) => {
          if (colDef.params.templateUrl && cellNode.length) {
            this.auUtilService
              .createAureliaViewModelAddToSlot(
                colDef.params.templateUrl,
                { model: dataContext.FieldA /** THIS BECOMES TYPE SAFE **/}, cellNode[0], true
             );
          }
        },
        params: {
          templateUrl: "path/to/my/template",
        },
        },
      ];
    }
}
```

With this sample the `field` prop in the columnDefinitions is typesafe and will scream on providing something different, as well as the `dataContext` for the asyncPostRenderer.

Now if we could initialize the Column interface with a generic used for the GridFields equivalent of above, we could also use the generic for the fields and dataContext as I did above manually.

What do you think? I know this is most likely a breaking change, but I really think the current use of the generic doesn't bring any value to the table, except I'm missing a point here.

**Comment by ghiscoding** [comment]
I was looking at adding more Types in my other lib (Slickgrid-Universal) and I did but I had a bit of a problem understanding how to fully add Type for the Column, I wanted to pass the user's column type as `Column<T>` when you define your column definitions and I asked a question about this on Stack Overflow [Problem with using a defined TypeScript Generic Type in an external function using the Type T](https://stackoverflow.com/questions/62220906/problem-with-using-a-defined-typescript-generic-type-in-an-external-function-usi) and this Slickgrid-Universal [Example](https://github.com/ghiscoding/slickgrid-universal/blob/master/examples/webpack-demo-vanilla-bundle/src/examples/example04.ts#L44) is the one that goes with the question. It might not be exactly what you're describing though but it's in the same context. The intention was to pass the data Type to the `Column<T>` and it would flow to a few other interfaces like the `asyncPostRender` on `dataContext: T`. 

However note that Aurelia-Slickgrid doesn't have full Type support for the core SlickGrid and DataView, they are created as [SlickGrid](https://github.com/ghiscoding/aurelia-slickgrid/blob/master/src/aurelia-slickgrid/models/slickGrid.interface.ts) and [SlickDataView](https://github.com/ghiscoding/aurelia-slickgrid/blob/master/src/aurelia-slickgrid/models/slickDataView.interface.ts) interfaces but they are optional, I added them long after creating the lib and making the lib fully types would break lot of people's code, so I made them as optional and all of my services it still refer to `grid: any` and `dataView: any`. The goal is to eventually remove all the core/common code from Aurelia-Slickgrid (and Angular-Slickgrid) and use Slickgrid-Universal common place and when that happen (in few months), I'll do a major/breaking version of Aurelia-Slickgrid with full type. 

To be fair I haven't used much of the `Column<T>`, it didn't fully work as I wanted (as discussed in the SO), so if you want to change that portion, we could still do it since SlickGrid & DataView are still optional in Aurelia-Slickgrid and barely any (if any) uses Slickgrid-Universal and there's no Type in Salesforce where I use it because it's ES6 syntax. I don't think anyone knows about the `Column<T>` since I didn't advertise it in any demo.

**Comment by zewa666** [comment]
Errors is a hefty word. I would say the ide cases merely differ. The primary need for typesafety for me is on the field property of the column since this gets mapped to a fetched result when populating data and refactoring string based identifiers is a major PITA. Granted though your approach could be used as well with casting field as `key in ReportItem`. Anyways I guess your approach to moving the core to universal is certainly a good one and as it stands for now as described initially I have my workaround with a few custom casts so no need to change in that case.

**Comment by ghiscoding** [comment]
I'm totally open to any contributions, since they are so rare I do enjoy them whenever they come. I didn't have good result with `Column<T>`, so I'm open to anything that could help improve the lib in any ways.

**Comment by zewa666** [comment]
You'll gonna like and hate what you see in a couple of minutes :)

**Comment by zewa666** [comment]
here it is. Its draft on purpose, you'll see why when reading it https://github.com/ghiscoding/aurelia-slickgrid/pull/442

**Comment by ghiscoding** [comment]
closed by PR #466
